### ISSUE TITLE
GPL-593 Migrate old samples collections to include their timestamps into the new single samples collection [m]

### ISSUE THREAD

**Issue by emrojo** [sanger/crawler#84]
**User story**
Low priority.

After completing GPL577 :

Previously to GPL577 samples were recreated and stored in a different collection every day. Now we will have all samples in same collection but we still want to know when the sample was first inserted in the database. To use it we want to apply the timestamp of the collection the sample firstly appeared into the new samples single collection.

**Who are the primary contacts for this story**
Andrew
Eduardo

**Acceptance criteria**
To be considered successful the solution must allow:
- [ ] All samples should have a timestamp with the date and time when the sample was inserted in the database

**Dependencies**
This story is blocked by the following dependencies:
- #<issue_no.>
- sanger/<repo>#<issue_no.>

**Additional context**

**Comment by KatyTaylor** [comment]
@emrojo - do you know the answer to this...

Looking at the collections (in prod), there are a few different formats of collection name. Examples below:
1. `samples` - the main samples collection (just one collection like this)
2. `samples_07052020_1251` (3 collections with date format like this)
3. `samples_200519_1415` (63 collections with date format like this)
4. `tmp_samples_200709_1240` (14 collections like this)

Should all of 2-4 be included in the migration, or are the 'tmp' ones irrelevant?

Thanks

**Comment by KatyTaylor** [comment]
Results from testing on UAT yesterday:

Summary:
* Seems to have worked fine
* Was interrupted part way through due to my computer going to sleep - I commented out the early bits of code that didn't need to be re-run, hardcoded the list of collections it still had to get through, and re-started it
* Took 7.5 hours
* Had a false start initially where I got the MongoDB 'document too large' issue, solved it with this commit - https://github.com/sanger/crawler/commit/d49ac2a807f3bb5040e8336dd3a7f65ade9fce0b

Verifying results:
Around 70k samples were not updated. Not sure why (maybe there is a missing file in UAT), but is only around 1% so probably OK. 
`psd_lims_rs:PRIMARY> db.samples.count({'migrated_temp': null})`
`70251`
`psd_lims_rs:PRIMARY> db.samples.count({'created_at': null})`
`70251`

Key fields from 5 records of `samples` collection, retrieved using $sample aggregation:
```
"concat_id" : "GLM00023901 - GLA002140R_H10 - Negative - GLS", 
"created_at" : ISODate("2020-06-11T10:24:00Z"), 
"migrated_temp" : "Yes"

"concat_id" : "SAT00772606 - AP-rna-00114435_F07 - Negative - AP", 
"created_at" : ISODate("2020-06-27T01:00:00Z"), 
"migrated_temp" : "Yes"

"concat_id" : "TIP00026142 -  - Void - AP", 
"created_at" : ISODate("2020-06-10T01:00:00Z"), 
"migrated_temp" : "Yes"

"concat_id" : "PBO00024522 - 002107575_C11 - Negative - CB", 
"created_at" : ISODate("2020-07-30T15:02:00Z"), 
"migrated_temp" : "Yes"

"concat_id" : "MCM00030385 - AP-rna-00113443_F09 - Negative - AP", 
"created_at" : ISODate("2020-06-10T01:00:00Z"), 
"migrated_temp" : "Yes"
```

Detailed log from command line output during migration:
[UAT_test_migration_2020_08_11.txt](https://github.com/sanger/crawler/files/5061916/UAT_test_migration_2020_08_11.txt)

**Comment by KatyTaylor** [comment]
Documentation of the prod migration - https://ssg-confluence.internal.sanger.ac.uk/pages/viewpage.action?pageId=111052537
